arduino-ide-extension/src/common/protocol/sketches-service.ts

```diff
@@ -151,6 +152,51 @@ export interface Sketch extends SketchRef {
   readonly rootFolderFileUris: string[]; // `RootFolderFiles` (does not include the main sketch file)
 }
 export namespace Sketch {
+  // (non-API) exported for the tests
+  export const invalidSketchFolderNameMessage = nls.localize(
+    'arduino/sketch/invalidSketchName',
+    'Sketch names must start with a letter or number, followed by letters, numbers, dashes, dots and underscores. Maximum length is 63 characters.'
+  );
+  const invalidCloudSketchFolderNameMessage = nls.localize(
+    'arduino/sketch/invalidCloudSketchName',
+    'The name must consist of basic letters, numbers, or underscores. The maximum length is 36 characters.'
+  );
+  /**
+   * `undefined` if the candidate sketch folder name is valid. Otherwise, the validation error message.
+   * Based on the [specs](https://arduino.github.io/arduino-cli/latest/sketch-specification/#sketch-folders-and-files).
+   */
+  export function validateSketchFolderName(
+    candidate: string
+  ): string | undefined {
+    return /^[0-9a-zA-Z]{1}[0-9a-zA-Z_\.-]{0,62}$/.test(candidate)
+      ? undefined
+      : invalidSketchFolderNameMessage;
+  }
+
+  /**
+   * `undefined` if the candidate cloud sketch folder name is valid. Otherwise, the validation error message.
+   * Based on how https://create.arduino.cc/editor/ works.
+   */
+  export function validateCloudSketchFolderName(
+    candidate: string
+  ): string | undefined {
+    return /^[0-9a-zA-Z_]{1,36}$/.test(candidate)
```

Comment:
```suggestion
    return /^[0-9a-zA-Z]{1}[0-9a-zA-Z_]{0,35}$/.test(candidate)
```

I'm ready to accept that there is some hypothetical reason why the maximum length must be more restrictive for Cloud sketch. At least in this case a valid cloud sketch name length is still in compliance with the specification, even though the converse is not true.

But it is unacceptable to allow Cloud sketch names that are in violation of the Arduino Sketch Specification.

Reply:
> But it is unacceptable to allow Cloud sketch names that are in violation of the Arduino Sketch Specification.

It was requested in one of the design files that IDE2 must rename sketches when cloning from and pushing to the cloud. We need to add tests to verify IDE2 can do the name transformation.

Users can still create and pull a remote sketch, so relaxing the regex here might not help. What do you think?

Reply:
>Users can still create and pull a remote sketch, so relaxing the regex here might not help. What do you think?

I don't think it is a problem because the new approach is to validate names when the sketch is opened. So even if an invalid sketch is pulled, it will still be at the time the user opens the pulled sketch.

I'm referring to this feature:

https://github.com/arduino/arduino-ide/pull/1821#issuecomment-1403416987

> - When you open an invalid sketch, IDE2 will prompt for user actions to fix the issues.
>   - First, IDE2 prompts for the sketch folder name fix as a `Save as...` operation. IDE2 prompt until the user specifies a valid folder name or aborts the operation.
>   - If the sketch folder name has been corrected, IDE2 reloads and IDE2 validates the code files
>   - If there are any invalid code files, the user has to fix them as a `Rename` operation.
>   - When all problems have been fixed, IDE2 is ready for use. If no, IDE2 prompts users whether they want to interrupt the operation and navigate to a new temp sketch.

Reply:
> I don't think it is a problem because the new approach is to validate names when the sketch is opened. So even if an invalid sketch is pulled, it will still be at the time the user opens the pulled sketch.

You're correct, but it only covers some user cases. After IDE2 pulls the cloud sketch and fixes any possible sketch folder name and code file name errors, the user can create a file remotely and pull it again. It will be spec incompatible if the name of the code file starts with an underscore. Please correct me if I am overlooking something. Thanks!

IDE2 may run a validation after pulling from the remote.

Reply:
>the user can create a file remotely and pull it again. It will be spec incompatible if the name of the code file starts with an underscore. Please correct me if I am overlooking something.

You are right that this would provide a path to having a non-compliant sketch open in Arduino IDE. However, that will only be possible during that session. The next time the user tries to open the sketch, the added file will be validated.

So I wouldn't consider this scenario to completely invalidate the proposed approach of only doing the filename validation on an open operation.

Ideally validation would also be done when a file is added or renamed through an "external" process such as a pull. There is an equivalent scenario for local sketches in that the user may add or rename files of the open sketch externally. But I don't consider such validation to be absolutely essential. The most important reason for validating against the specification is to ensure shared sketches are compatible with any official Arduino development tool as well as 3rd party tools from conscientious developers. I think it would be rare that a developer would not end up later opening the type of sketch that is shared (as opposed to a quick throwaway test sketch we might write and use during a single IDE session and never look at again).

So if it is relatively easy to add a validation triggered by external addition or name change of sketch file, then great. But if this is something of significant difficulty to implement then I think the "on open" approach is sufficient.